shubhrajps/levels-fyi-countries
# [MAJOR] Fetch countries information

# Added feature to fetch countries information
Deployed at https://levels-fyi-countries.vercel.app/

## Changes Made

1. New Components:
- CountryExplorer: The main component that handles the fetching and display of country data, including a search feature for filtering countries.
- CountryCard: Displays individual country information, including flags, population, capital, and more.
- Skeleton: A loading placeholder component that shows while the country data is being fetched.

2. API Route:
- /api/countries : an API call using Axios to fetch country data from https://restcountries.com/. Added error handling for the API request to manage potential failures gracefully.

3. Responsive Design:
- Implemented responsive behavior in the CountryExplorer component for mobile and desktop views based on the window size.

4. Testing:
- Added unit tests for all components using Jest and React Testing Library.
- Created integration tests for the API route to ensure it returns the correct data.

5. CI/CD Pipeline:
- Set up a GitHub Actions workflow to run tests and deploy to Vercel upon pushing changes to the main branch or creating a pull request.

## Screenshots & Videos

### Homescreen
![Screenshot 2024-09-09 at 7 10 50 PM](https://github.com/user-attachments/assets/59ba6dc5-f950-4bd6-9821-be34023bd198)

### Skeleton loading
![Screenshot 2024-09-09 at 7 10 21 PM](https://github.com/user-attachments/assets/412f2ab0-8934-447d-b8bd-3afee3fc32df)

### Search Functionality
![Screenshot 2024-09-09 at 7 11 00 PM](https://github.com/user-attachments/assets/99c38615-f883-45ce-ab39-c9543934f35d)

### Flip to view more details
![Screenshot 2024-09-09 at 7 11 56 PM](https://github.com/user-attachments/assets/ce13c5db-89c4-4b30-84cf-8e89003ed7a8)

### Live Video
https://github.com/user-attachments/assets/66b3f991-f323-4021-aa86-a76b59eabcb6

## How to test
1. Clone the repository `git clone https://github.com/shubhrajps/levels-fyi-countries.git`
2. Check the branch `git checkout fetch-countries-info`
3. Install the dependencies `npm install`
4. Run the project `npm run dev`
5. Access the project at `https://localhost:3000`

## Checklist
- [x]  Tests have been written and pass successfully.
- [x]  CI/CD pipeline is configured and working.
- [x]  Changes have been tested in both mobile and desktop views.
